File: scratch/record-scalarization-integration-20260916/step1/src/Test/LazyEvaluationFFI.go
package Test_LazyEvaluationFFI


type Lazy func() int

func force(l Lazy) int {
	return l()
}

func deferFunc(f func() int) Lazy {
	return f
}

func buildThunks(depth int, acc Lazy) Lazy {
	if depth == 0 {
		return acc
	}
	return buildThunks(depth-1, deferFunc(func() int {
		return force(acc) + 1
	}))
}

func runManyTimes(times int, acc int) int {
	if times == 0 {
		return acc
	}
	return runManyTimes(times-1, acc+force(buildThunks(1000, deferFunc(func() int {
		return 0
	}))))
}

func RunLazyEvaluationFFI(limit int) int {
	n := int(limit)
	return (runManyTimes(n, 0))
}

File: scratch/record-scalarization-integration-20260916/step1/src/Test/LazyEvaluationFFI.rs
use std::rc::Rc;

type Lazy<'a> = Rc<dyn Fn() -> i64 + 'a>;

fn force(l: &Lazy) -> i64 {
    l()
}

fn defer<'a, F: Fn() -> i64 + 'a>(f: F) -> Lazy<'a> {
    Rc::new(f)
}

fn build_thunks<'a>(depth: i64, acc: Lazy<'a>) -> Lazy<'a> {
    if depth == 0 {
        return acc;
    }
    let prev = acc.clone();
    build_thunks(depth - 1, defer(move || force(&prev) + 1))
}

pub fn Test_LazyEvaluationFFI_runLazyEvaluationFFI(mut limit: i64) -> i64 {
    let mut acc = 0;
    for _ in 0..limit {
        let t = build_thunks(1000, defer(|| 0));
        acc += force(&t);
    }
    acc
}
